pkg/controller/jobframework/reconciler.go

```diff
@@ -572,68 +569,94 @@ func (r *JobReconciler) recordAdmissionCheckUpdate(wl *kueue.Workload, job Gener
-func (r *JobReconciler) getAncestorJobManagedByKueue(ctx context.Context, jobObj client.Object) (client.Object, error) {
-	seen := sets.New[types.UID]()
-	currentJob := jobObj
+func FindAncestorJobManagedByKueue(managedOwnersChain []client.Object) client.Object {
+	for i := len(managedOwnersChain) - 1; i >= 0; i-- {
+		owner := managedOwnersChain[i]
+		if QueueNameForObject(owner) != "" {
+			return owner
+		}
+	}
+	return nil
+}
+
+func ManagedOwnersChain(ctx context.Context, c client.Client, record record.EventRecorder, jobObj client.Object) ([]client.Object, error) {
+	var (
+		chain      = make([]client.Object, 0, managedOwnersChainLimit)
+		seen       = sets.New[types.UID]()
+		currentObj = jobObj
+	)
+
 	for {
-		if seen.Has(currentJob.GetUID()) {
-			return nil, nil
+		if seen.Has(currentObj.GetUID()) {
+			break
 		}
-		seen.Insert(currentJob.GetUID())
+		seen.Insert(currentObj.GetUID())
 
-		owner := metav1.GetControllerOf(currentJob)
-		if owner == nil || !IsOwnerManagedByKueue(owner) {
-			return nil, nil
+		owner := metav1.GetControllerOf(currentObj)
+		if owner == nil {
+			break
 		}
-		parentJob := GetEmptyOwnerObject(owner)
-		if parentJob == nil {
-			return nil, fmt.Errorf("workload owner %v: %w", owner, ErrUnknownWorkloadOwner)
+
+		managed := true
+
+		parentObj := GetEmptyOwnerObject(owner)
+		if parentObj == nil {
+			managed = false
+			parentObj = &metav1.PartialObjectMetadata{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: owner.APIVersion,
+					Kind:       owner.Kind,
+				},
+			}
 		}
-		if err := r.client.Get(ctx, client.ObjectKey{Name: owner.Name, Namespace: jobObj.GetNamespace()}, parentJob); err != nil {
+
+		if err := c.Get(ctx, client.ObjectKey{Name: owner.Name, Namespace: jobObj.GetNamespace()}, parentObj); err != nil {
 			return nil, errors.Join(ErrWorkloadOwnerNotFound, err)
 		}
-		if QueueNameForObject(parentJob) != "" {
-			return parentJob, nil
+
+		// Ignore unmanaged resources and disabled jobs.
+		if managed {
+			chain = append(chain, parentObj)
 		}
-		currentJob = parentJob
+
+		currentObj = parentObj
 		if len(seen) > managedOwnersChainLimit {
-			r.record.Eventf(jobObj, corev1.EventTypeWarning, ReasonJobNestingTooDeep,
-				"Terminated search for Kueue-managed Job because ancestor depth exceeded limit of %d", managedOwnersChainLimit)
+			record.Eventf(jobObj, corev1.EventTypeWarning, ReasonJobNestingTooDeep,
+				"Terminated search for Kueue-managed Job because ancestor depth exceeded limit of %d", managedOwnersChainLimit,
+			)
 			ctrl.LoggerFrom(ctx).V(2).Info(
-				"Terminated search for Kueue-managed Job because ancestor depth exceeded managedOwnersChainlimit",
+				"WARNING: Terminated search for Kueue-managed Job because ancestor depth exceeded managedOwnersChainLimit",
 				"limit ", managedOwnersChainLimit,
-				"lastParentReached", parentJob,
+				"lastParentReached", parentObj,
 			)
-			return nil, nil
+			break
```

Comment:
I would suggest to just keep the `return` for the readability, and smaller diff.